borsh-derive-internal/src/enum_de.rs

```diff
@@ -70,9 +85,15 @@ pub fn enum_de(input: &ItemEnum, cratename: Ident) -> syn::Result<TokenStream2>
             }
             Fields::Unit => {}
         }
-        variant_arms.extend(quote! {
-            if variant_tag == #discriminant { #name::#variant_ident #variant_header } else
-        });
+        if use_discriminant {
+            variant_arms.extend(quote! {
+                if variant_tag == #discriminant { #name::#variant_ident #variant_header } else
```

Comment:
> Why aren’t we doing `#discriminant => #name::#variant_ident #variant_header,` here as well? It seems it would make the rest of the code work without change if in both cases we used a match.

I've added following tests to `enum_de.rs`:
```rust
    #[test]
    fn borsh_discriminant_false() {
        let item_enum: ItemEnum = syn::parse2(quote! {

            enum X {
                A,
                B = 20,
                C,
                D,
                E = 10,
                F,
            }
        })
        .unwrap();
        let actual = enum_de(
            &item_enum,
            Ident::new("borsh", Span::call_site()),
            Some(false),
        )
        .unwrap();

        insta::assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }
    #[test]
    fn borsh_discriminant_true() {
        let item_enum: ItemEnum = syn::parse2(quote! {

            enum X {
                A,
                B = 20,
                C,
                D,
                E = 10,
                F,
            }
        })
        .unwrap();
        let actual = enum_de(
            &item_enum,
            Ident::new("borsh", Span::call_site()),
            Some(true),
        )
        .unwrap();

        insta::assert_snapshot!(pretty_print_syn_str(&actual).unwrap());
    }

```
These produce following expansions:
```rust
// false
impl borsh::de::EnumExt for X {
    fn deserialize_variant<R: borsh::__private::maybestd::io::Read>(
        reader: &mut R,
        variant_idx: u8,
    ) -> ::core::result::Result<Self, borsh::__private::maybestd::io::Error> {
        let mut return_value = match variant_idx {
            0u8 => X::A,
            1u8 => X::B,
            2u8 => X::C,
            3u8 => X::D,
            4u8 => X::E,
            5u8 => X::F,
            _ => {
                return Err(
                    borsh::__private::maybestd::io::Error::new(
                        borsh::__private::maybestd::io::ErrorKind::InvalidInput,
                        borsh::__private::maybestd::format!(
                            "Unexpected variant index: {:?}", variant_idx
                        ),
                    ),
                );
            }
        };
        Ok(return_value)
    }
}
```
```rust
// true
impl borsh::de::EnumExt for X {
    fn deserialize_variant<R: borsh::__private::maybestd::io::Read>(
        reader: &mut R,
        variant_tag: u8,
    ) -> ::core::result::Result<Self, borsh::__private::maybestd::io::Error> {
        let mut return_value = if variant_tag == 0 {
            X::A
        } else if variant_tag == 20 {
            X::B
        } else if variant_tag == 20 + 1 {
            X::C
        } else if variant_tag == 20 + 1 + 1 {
            X::D
        } else if variant_tag == 10 {
            X::E
        } else if variant_tag == 10 + 1 {
            X::F
        } else {
            return Err(
                borsh::__private::maybestd::io::Error::new(
                    borsh::__private::maybestd::io::ErrorKind::InvalidInput,
                    borsh::__private::maybestd::format!(
                        "Unexpected variant tag: {:?}", variant_tag
                    ),
                ),
            )
        };
        Ok(return_value)
    }
}
```
As @mina86  pointed out, i can see no reason to use `match` in one case and `if-else_if-else` chain in the other.
`match` in both cases should be fine.
Also, there's no reason to use `variant_tag` for name of method argument in one case, and `variant_idx` - in the other.
If these small discrepancies are removed, the code for both variants will be a lot easier to be unified and read.

Reply:
Ok, rust doesn't allow to use expressions at the spot, where a pattern is expected, so it should either be:
1. use `if-else_if-else` chain in both cases
2. or use `match` in both cases with [evalexpr crate](https://crates.io/crates/evalexpr), which work fine for integer literals in place of enum discriminants (but not other arbitrary expressions), similar to the following:
```diff
diff --git a/borsh-derive-internal/src/enum_discriminant_map.rs b/borsh-derive-internal/src/enum_discriminant_map.rs
index 9d7545a1..66f41ce0 100644
--- a/borsh-derive-internal/src/enum_discriminant_map.rs
+++ b/borsh-derive-internal/src/enum_discriminant_map.rs
@@ -1,24 +1,30 @@
 use std::collections::HashMap;
 
+use evalexpr::*;
+
 use proc_macro2::{Ident, TokenStream};
 use quote::quote;
 use syn::{punctuated::Punctuated, token::Comma, Variant};
 
 /// Calculates the discriminant that will be assigned by the compiler.
 /// See: https://doc.rust-lang.org/reference/items/enumerations.html#assigning-discriminant-values
-pub fn discriminant_map(variants: &Punctuated<Variant, Comma>) -> HashMap<Ident, TokenStream> {
+pub fn discriminant_map(variants: &Punctuated<Variant, Comma>) -> Result<HashMap<Ident, u8>, EvalexprError> {
     let mut map = HashMap::new();
 
     let mut next_discriminant_if_not_specified = quote! {0};
 
     for variant in variants {
         let this_discriminant = variant.discriminant.clone().map_or_else(
             || quote! { #next_discriminant_if_not_specified },
             |(_, e)| quote! { #e },
         );
         next_discriminant_if_not_specified = quote! { #this_discriminant + 1 };
-        map.insert(variant.ident.clone(), this_discriminant);
+
+        let value = eval(&this_discriminant.to_string())?;
+        let value_int = value.as_int()?;
+        
+        map.insert(variant.ident.clone(), value_int as u8);
     }
 
-    map
+    Ok(map)
 }
```